pyanaconda/modules/boss/boss.py

```diff
@@ -0,0 +1,161 @@
+        self._started_module_services = []
+        self._failed_module_services = []
+        self._loop = GLib.MainLoop()
+        self._bus = get_bus()
+        self._dbus = self._bus.get('org.freedesktop.DBus', '/org/freedesktop/DBus')
+
+    @property
+    def modules(self):
+        return self._modules
+
+    @property
+    def addons(self):
+        return self._addons
+
+    @property
+    def required_module_services(self):
+        services = ["%s.%s" % (constants.DBUS_MODULE_NAMESPACE, m)
+                    for m in self.modules]
+        services.extend(["%s.%s" % (constants.DBUS_ADDON_NAMESPACE, a)
+                         for a in self.addons])
+        return services
+
+    @property
+    def running_module_services(self):
+        # For our purpose, just this
+        return self._started_module_services
+
+    def EchoString(self, s: Str) -> Str:
+        """returns whatever is passed to it"""
+        log.debug(s)
+        return s
+
+    def Quit(self):
+        """removes the object from the DBUS connection and exits"""
+        log.debug("sending Quit to all modules and addons")
+        for service in self.running_module_services:
+            module = self._bus.get(service)
+            # TODO: async
+            try:
+                module.Quit()
+            except Exception:
+                log.exception("Boss shutdown failed")
+        self._loop.quit()
+
+    def modules_starting_finished(self):
+        # TODO: do we need some synchronization here?
+        return set(self._started_module_services + self._failed_module_services) == set(self.required_module_services)
+
+    def _start_service_cb(self, service, returned=None, error=None):
+        """callback for dbus.StartServiceByName"""
+        if error:
+            log.debug("%s error: %s" % (service,error))
+            self._failed_module_services.append(service)
+        elif returned:
+            if returned == DBUS_START_REPLY_SUCCESS:
+                log.debug("%s started successfully, returned: %s)" % (service, returned))
+                self._started_module_services.append(service)
+            else:
+                log.debug("%s failed to start, returned: %s" % (service, returned))
+                self._failed_module_services.append(service)
+
+        self.check_modules_started()
+
+    def check_modules_started(self):
+        if self.modules_starting_finished():
+            log.debug("modules starting finished, running: %s failed: %s" %
+                      (self._started_module_services,
+                       self._failed_module_services))
+            for service in self._started_module_services:
+                log.debug("SERVICE: %s", service)
+                module = self._bus.get(service)
+                module.EchoString("Boss told me that all required modules were started: %s or failed: %s." %
+                                  (self._started_module_services, self._failed_module_services))
```

Comment:
Do you mean the EchoString call ? That's basically for testing/rough demonstration that modules & inter module communication works. It would certainly not be done once the modules actually do something. But I guess we might keep the EchoString function or something similar for simple testing if modules are running & in reasonable state.